src/kernels/composable_kernel/include/kernel_algorithm/gridwise_generic_2d_reduction_blockwise.hpp

```diff
@@ -0,0 +1,462 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef CK_GRIDWISE_GENERIC_2D_REDUCTION_BLOCKWISE_HPP
+#define CK_GRIDWISE_GENERIC_2D_REDUCTION_BLOCKWISE_HPP
+
+#include "float_type.hpp"
+#include "reduction_operator.hpp"
+#include "reduction_functions.hpp"
+#include "reduction_common.hpp"
+
+#include "blockwise_generic_tensor_slice_copy.hpp"
+#include "ConstantMatrixDescriptor.hpp"
+
+namespace ck {
+
+template <int BlockSize,
```

Comment:
Ok

Reply:
I prefer to have those integers which are not calculated with the indexes/lengths of Sequence<> using "int",   because someday, "index_t" could be defined to "long/unsigned long",   but for reduction "indices", the type is  always "int" and it is specified by the miopenReduceTensor() api.

Reply:
You don't need to use ```index_t``` for reduction "indices" which is part of your reduction result, see my other comment below. But please use ```index_t``` for everything else.

```index_t``` is used for address calculation inside tensor descriptor, and since every integer variable a developer defines will eventually be used in these address calculation, composable kernel controls what type is used, otherwise bad performing ISA may be created

If you want to control the type of  "reduction indices", please use ```static_cast```, between it and ```index_t```. Sorry for this kind of mess.

Reply:
Check [https://github.com/ROCmSoftwarePlatform/MIOpen/pull/322/commits/0802a88a7bd885e8ee022fcafc371104e5514b4c]

Reply:
@asroy Has this been completed?